lldb/examples/python/parsed_cmd.py

```diff
@@ -0,0 +1,315 @@
+"""
+This module implements a couple of utility classes to make writing
+lldb parsed commands more Pythonic.
+The way to use it is to make a class for you command that inherits from ParsedCommandBase.
+That will make an LLDBOVParser which you will use for your
+option definition, and to fetch option values for the current invocation
+of your command.  Access to the OV parser is through:
+
+ParsedCommandBase.get_parser()
+
+Next, implement setup_command_definition in your new command class, and call:
+
+  self.get_parser().add_option
+
+to add all your options.  The order doesn't matter for options, lldb will sort them
+alphabetically for you when it prints help.
+
+Similarly you can define the arguments with:
+
+  self.get_parser.add_argument
+
+at present, lldb doesn't do as much work as it should verifying arguments, it pretty
+much only checks that commands that take no arguments don't get passed arguments.
+
+Then implement the execute function for your command as:
+
+    def __call__(self, debugger, args_array, exe_ctx, result):
+
+The arguments will be in a python array as strings.  
+
+You can access the option values using varname you passed in when defining the option.  
+If you need to know whether a given option was set by the user or not, you can retrieve 
+the option definition array with:
+
+  self.get_options_definition()
+
+look up your element by varname and check the "_value_set" element.
+
+There are example commands in the lldb testsuite at:
+
+llvm-project/lldb/test/API/commands/command/script/add/test_commands.py
+
+FIXME: I should make a convenient wrapper for that. 
+"""
+import inspect
+import lldb
+import sys
+
+class LLDBOVParser:
+    def __init__(self):
+        self.options_array = []
+        self.args_array = []
+
+    # Some methods to translate common value types.  Should return a
+    # tuple of the value and an error value (True => error) if the
+    # type can't be converted.
+    # FIXME: Need a way to push the conversion error string back to lldb.
```

Comment:
Do we raise Python exceptions anywhere else in our API's.  It is a pythonic way to do things, but we don't tend to do it that way.

Reply:
We don't do it often but there is prior art. Specifically in a few of our SBAPI extensions, such as SBAddress and SBData. My $0.02 is that it's worth making these interfaces more pythonic since people will want to write their own commands with python.

Reply:
I still think exceptions would be an inconvenient way to do this.  These aren't API's that users are going to call directly, so we aren't matching expectations for Python function behavior.  Rather, in the middle of parsing a command from lldb (happening in C++) the command parser will arrange to have this called.  So if it raises a Python exception, the LLDB SWIG wrappers will have to catch that exception and turn it into a Status with the error we want to report.  I think it would be much less fragile to just pipe an SBError through directly and use that for error reporting.